File: JavaScript/array/searchMatrix.js
/**
 * 將右上角或者左下角作為起點(start)來找。假如以右上角作為起點，那麼需要考慮三種情況：
 *  - 如果 target > start: 則橫坐標加一；
 *  - 如果 target < start: 則縱坐標減一；
 *  - 如果 target = start: 則返回 true。
 * 
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
  if (matrix === null || matrix.length === 0) {
    return false;
  }
  var row = matrix.length;
  var col = matrix[0].length;
  
  var i = 0;
  var j = col - 1;
  
  while (i <= row - 1 && j >= 0) {
    if (matrix[i][j] < target) {
      i = i + 1;
    } else if (matrix[i][j] > target) {
      j = j - 1;
    } else {
      return true;
    }
  }
  return false;
};

File: Python/array/searchMatrix.py
# -*- coding:UTF-8 -*-
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        @ Solution：
        - 可以从右上角或者左下角开始找，以右上角为例：（temp_i，temp_j是当前所在的位置）
            - 如果target大于当前值，要往下找，行加1；
            - 如果target小于当前值，要往左找，列减1；
            - 相等则返回TRUE；
        """
        if not matrix:
            return False
        row, col = len(matrix), len(matrix[0])
        temp_i, temp_j = 0, col-1
        while 0 <= temp_i < row and 0 <= temp_j < col:
            if target > matrix[temp_i][temp_j]:
                temp_i += 1
            elif target < matrix[temp_i][temp_j]:
                temp_j -= 1
            else:
                return True
        return False